File: apps/api/lib/observability/sentry-alerts.py
"""
apps/api/lib/observability/sentry-alerts.py — Sentry alert wiring (backend).

1st release launch (cj-style 64번째 진입점) — T5.3 (AC #5.4) — F18.5 Production verification.
- Sentry alert wiring backend production 환경 결정 wire.
- 5 alert rules: 5xx / tenant isolation / alembic / audit log / PITR drill.
"""
from __future__ import annotations

import logging
from typing import Final

logger = logging.getLogger(__name__)

# 5 alert rules (cj-style 1st release launch 64번째 진입점 정합)
ALERT_RULES: Final[tuple[str, ...]] = (
    "5xx_api_error_rate",
    "tenant_isolation_violation",
    "alembic_migration_failure",
    "audit_log_integrity_failure",
    "pitr_drill_overdue",
)


def capture_alert(rule: str, message: str, *, level: str = "error", tags: dict | None = None, extra: dict | None = None) -> None:
    """Capture a Sentry alert for a production environment violation.

    1st release launch 결정 wire — production environment alert wiring.
    """
    try:
        import sentry_sdk

        sentry_sdk.capture_message(
            f"[1st-release] {rule}: {message}",
            level=level,
            tags={"alert_rule": rule, **(tags or {})},
            extra=extra,
        )
    except ImportError:
        logger.warning("sentry-sdk not installed; alert not sent: %s — %s", rule, message)


def check_tenant_isolation(violation: bool, tenant_id: str) -> None:
    """Check tenant isolation (CR 0-2 RLS lesson) and capture alert if violated."""
    if violation:
        capture_alert(
            "tenant_isolation_violation",
            f"Tenant isolation violation detected for tenant_id={tenant_id}",
            level="fatal",
            tags={"tenant_id": tenant_id},
        )

File: apps/web/lib/observability/sentry-alerts.ts
/**
 * apps/web/lib/observability/sentry-alerts.ts — Sentry alert wiring (frontend).
 *
 * 1st release launch (cj-style 64번째 진입점) — T5.3 (AC #5.3) — F18.5 Production verification.
 * - Sentry alert wiring production 환경 결정 wire.
 * - 5 alert rules: 5xx / auth error / LISTEN drop / backup fail / 2FA fail.
 */
import * as Sentry from "@sentry/nextjs";

export type AlertRule =
  | "5xx_error_rate"
  | "auth_error_rate"
  | "listen_notify_connection_drop"
  | "backup_failure"
  | "two_fa_verification_failure";

export interface AlertPayload {
  rule: AlertRule;
  message: string;
  level: "warning" | "error" | "fatal";
  tags?: Record<string, string>;
  extra?: Record<string, unknown>;
}

export function captureAlert(payload: AlertPayload): void {
  Sentry.captureMessage(`[1st-release] ${payload.rule}: ${payload.message}`, {
    level: payload.level,
    tags: { alert_rule: payload.rule, ...(payload.tags ?? {}) },
    extra: payload.extra,
  });
}

export const ALERT_THRESHOLDS: Readonly<Record<AlertRule, { threshold: number; windowSec: number }>> = {
  "5xx_error_rate": { threshold: 0.01, windowSec: 300 },
  auth_error_rate: { threshold: 0.05, windowSec: 300 },
  listen_notify_connection_drop: { threshold: 1, windowSec: 60 },
  backup_failure: { threshold: 1, windowSec: 86400 },
  two_fa_verification_failure: { threshold: 0.1, windowSec: 600 },
};
